src/mlpack/tests/ann/dag_network_test.cpp

```diff
@@ -0,0 +1,159 @@
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/ann.hpp>
+
+#include "../catch.hpp"
+
+using namespace mlpack;
+
+TEST_CASE("DAGNetworkCheckEmptyGraphTest", "[DAGNetworkTest]")
+{
+  DAGNetwork<NegativeLogLikelihood, RandomInitialization> model;
+  model.InputDimensions() = { 1, 2, 3 };
+
+  arma::mat testInput = arma::ones(6);
+  arma::mat testOutput;
+
+  REQUIRE_THROWS_AS(model.Predict(testInput, testOutput), std::invalid_argument);
+}
+
+TEST_CASE("DAGNetworkCheckCycleTest", "[DAGNetworkTest]")
+{
+  DAGNetwork<NegativeLogLikelihood, RandomInitialization> model;
+  model.InputDimensions() = { 1, 2, 3 };
+
+  size_t layer0 = model.Add<Linear>(10);
+  size_t layer1 = model.Add<Linear>(10);
+  size_t layer2 = model.Add<Linear>(10);
+  size_t layer3 = model.Add<Linear>(10);
+
+  model.Connect(layer0, layer1);
+  model.Connect(layer1, layer2);
+  model.Connect(layer2, layer3);
+  model.Connect(layer3, layer0);
+
+  arma::mat testInput = arma::ones(6);
+  arma::mat testOutput;
+
+  REQUIRE_THROWS_AS(model.Predict(testInput, testOutput), std::logic_error);
+}
+
+TEST_CASE("DAGNetworkMultipleOutputsTest", "[DAGNetworkTest]")
+{
+  DAGNetwork<NegativeLogLikelihood, RandomInitialization> model;
+  model.InputDimensions() = { 1, 2, 3 };
+
+  size_t layer0 = model.Add<Linear>(10);
+  size_t layer1 = model.Add<Linear>(10);
+  size_t layer2 = model.Add<Linear>(10);
+  size_t layer3 = model.Add<Linear>(10);
+
+  model.Connect(layer1, layer3);
+  model.Connect(layer0, layer1);
+  model.Connect(layer1, layer2);
+
+  arma::mat testInput = arma::ones(6);
+  arma::mat testOutput;
+
+  REQUIRE_THROWS_AS(model.Predict(testInput, testOutput), std::logic_error);
+}
+
+TEST_CASE("DAGNetworkMultipleInputsTest", "[DAGNetworkTest]")
+{
+  DAGNetwork<NegativeLogLikelihood, RandomInitialization> model;
+  model.InputDimensions() = { 1, 2, 3 };
+
+  size_t layer0 = model.Add<Linear>(10);
+  size_t layer1 = model.Add<Linear>(10);
+  size_t layer2 = model.Add<Linear>(10);
+  size_t layer3 = model.Add<Linear>(10);
+
+  model.Connect(layer0, layer2);
+  model.Connect(layer1, layer2);
+  model.Connect(layer2, layer3);
+
+  arma::mat testInput = arma::ones(6);
+  arma::mat testOutput;
+
+  REQUIRE_THROWS_AS(model.Predict(testInput, testOutput), std::logic_error);
+}
+
+TEST_CASE("DAGNetworkForestTest", "[DAGNetworkTest]")
+{
+  DAGNetwork<NegativeLogLikelihood, RandomInitialization> model;
+  model.InputDimensions() = { 1, 2, 3 };
+
+  size_t layer0 = model.Add<Linear>(10);
+  size_t layer1 = model.Add<Linear>(10);
+  size_t layer2 = model.Add<Linear>(10);
+
+  size_t layer3 = model.Add<Linear>(10);
+  size_t layer4 = model.Add<Linear>(10);
+  size_t layer5 = model.Add<Linear>(10);
+
+  model.Connect(layer0, layer1);
+  model.Connect(layer1, layer2);
+
+  model.Connect(layer3, layer4);
+  model.Connect(layer4, layer5);
+
+  arma::mat testInput = arma::ones(6);
+  arma::mat testOutput;
+
+  REQUIRE_THROWS_AS(model.Predict(testInput, testOutput), std::logic_error);
+}
```

Comment:
These are really nice and concise "invalid network" tests. I can't really think of any more. Maybe a really long cycle or something... but I'm fairly sure your code would catch it either way.